chain/events/events_test.go

```diff
@@ -1334,20 +1264,15 @@ func TestStateChangedTimeout(t *testing.T) {
 
 	// with check func reporting done
 
-	fcs = &fakeCS{
-		t: t,
-		h: 1,
-
-		msgs:       map[cid.Cid]fakeMsg{},
-		blkMsgs:    map[cid.Cid]cid.Cid{},
-		callNumber: map[string]int{},
-		tsc:        newTSCache(2*build.ForkLengthThreshold, nil),
-	}
-	require.NoError(t, fcs.tsc.add(fcs.makeTs(t, nil, 1, dummyCid)))
+	fcs = newFakeCS(t)
+	events, err = NewEvents(context.Background(), fcs)
+	require.NoError(t, err)
 
-	events = NewEvents(context.Background(), fcs)
+	// XXX: Needed to set the latest head so "check" succeeds". Is that OK? Or do we expect
+	// check to work _before_ we've received any events.
+	fcs.advance(0, 1, nil)
```

Comment:
@magik6k is this OK?

Previously: if we hadn't yet seen a block, we'd ask for the latest head and call check on that.
Now: if we haven't yet seen a block, we don't call check.

I made this change because there was no guarantee that the caller would see events for blocks between that latest head and the new head.

Possible solutions:

1. Leave this as it is.
2. Provide a way to get the latest head that guarantees that the caller will also get all events after that head.

Reply:
I think this is ok, the main point of the check method is providing atomicity guarantees (state change can't be missed).

(though we should check that nothing depends on the check being synchronous)

Reply:
So, currently it _won't_ call the check function (at all) if no events have been received. I'll fix that.

Reply:
Ok, I've now made registering an observer return the current head. That should make this atomic and avoid any potential deadlocks.